ProjFS.Mac/PrjFSKextTests/KextAssertIntegration.m

```diff
@@ -0,0 +1,90 @@
+#import "KextAssertIntegration.h"
+#include <pthread.h>
+
+extern void Assert(const char* file, unsigned line, const char* expression);
+
+static pthread_key_t s_currentTestCaseKey;
+
+static PFSKextTestCase* CurrentTestCase()
+{
+    return (__bridge PFSKextTestCase*)pthread_getspecific(s_currentTestCaseKey);
+}
+
+@implementation PFSKextTestCase
+{
+    uint32_t expectedKextAssertionFailures;
+}
+
++ (void) initialize
+{
+    pthread_key_create(&s_currentTestCaseKey, NULL);
+}
+
+- (void) setUp
+{
+    pthread_setspecific(s_currentTestCaseKey, (__bridge void*)self);
+}
+
+- (void) tearDown
+{
+    XCTAssertEqual(self->expectedKextAssertionFailures, 0, "Expected assertions apparently did not occur");
+    self->expectedKextAssertionFailures = 0;
+    pthread_setspecific(s_currentTestCaseKey, NULL);
+}
+
+- (void) setExpectedFailedKextAssertionCount:(uint32_t)failedAssertionCount
+{
+    self->expectedKextAssertionFailures = failedAssertionCount;
+}
+
+- (bool) isKextAssertionFailureExpected
+{
+    if (self->expectedKextAssertionFailures > 0)
+    {
+        --self->expectedKextAssertionFailures;
+        return true;
+    }
+    else
+    {
+        return false;
+    }
+}
+
+@end
+
+// catches kext assert()
```

Comment:
Does `Assert` get wired up to catch `assert()`?

Reply:
[Yes, in the kernel, `assert()` is implemented as a macro that expands to calling the `Assert()` function.](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/kern/assert.h#L91) The macro is needed to make the filename, line number, and textual representation of the expression that unexpectedly evaluated to false appear in the error message, generating the actual error message and panic are implemented in `Assert()`. Implementing our own version of that function means we can do what we want with that information.

Reply:
[Yes, in the kernel, `assert()` is implemented as a macro that expands to calling the `Assert()` function.](https://github.com/apple/darwin-xnu/blob/a449c6a3b8014d9406c2ddbdc81795da24aa7443/osfmk/kern/assert.h#L91) The macro is needed to make the filename, line number, and textual representation of the expression that unexpectedly evaluated to false appear in the error message, generating the actual error message and panic are implemented in `Assert()`. Implementing our own version of that function means we can do what we want with that information.

Reply:
I was thinking this exact question, thanks for asking it @wilbaker!